subprojects/core-api/src/main/java/org/gradle/api/attributes/AttributeContainer.java

```diff
@@ -76,11 +75,14 @@ public interface AttributeContainer extends HasAttributes {
     <T> AttributeContainer attributeProvider(Attribute<T> key, Provider<? extends T> provider);
 
     /**
-     * Returns the value of an attribute found in this container, or <code>null</code> if
+     * Returns the value of an attribute found in this container, or {@code null} if
      * this container doesn't have it.
+     * <p>
+     * Supplying a {@code null} argument is deprecated and will return {@code null}.
+     *
      * @param <T> the type of the attribute
-     * @param key the attribute key
-     * @return the attribute value, or null if not found
+     * @param key the attribute key (should not be {@code null})
```

Comment:
I do not think this is necessary.

Yes, there are existing use cases where build logic is violating the nullity of this method.

But, the Gradle interface is non-null unless otherwise stated. If we followed this pattern everywhere we would pollute the public API javadoc with nullity noise.

For the sake of consistency, I would vote against stating this explicitly.

Reply:
I'm completely ambivalent here.  I thought it might be nice to note this explicitly for this case, so that if this bug fix (checking for null and failing fast) breaks anyone they might understand why faster.  

But I can remove it if you think it's too much noise.  I'll leave it to you to make the final call.

Reply:
I don't think its the end of the world if we include it, but I think its redundant. 

I think the deprecation warning is enough IMO

Reply:
Does this show up as non-null for Kotlin?  if it doesn't, we can't change it in 8.x, but we could consider changing it in 9.0.

I'm OK with the extra docs.  I would remove the detail from the `@param` since it's covered in the body.

Reply:
This should show as non-null for Kotlin.  I don't think there is a problem there, as I haven't changed the signature.